File: web/protocol/src/messages.ts
import type { HostConfig } from "./config.js";
import type {
  CaptureDevices,
  CaptureSelection,
  CaptureState,
  ParticipantRoomView,
  RoomState,
  StreamProcessingFlags,
} from "./types.js";

export interface WelcomeMsg {
  type: "welcome";
  clientId: string;
  view: ParticipantRoomView;
}

export interface ParticipantViewMsg {
  type: "participant-view";
  view: ParticipantRoomView;
}

export interface JoinMsg {
  type: "join";
  name: string;
  hasVideo: boolean;
  hasAudio: boolean;
  clientId?: string;
}

export interface LeaveMsg {
  type: "leave";
}

export interface OfferMsg {
  type: "offer";
  from: string;
  to: string;
  sdp: string;
}

export interface AnswerMsg {
  type: "answer";
  from: string;
  to: string;
  sdp: string;
}

export interface ICECandidateMsg {
  type: "ice-candidate";
  from: string;
  to: string;
  candidate: string;
}

export interface ErrorMsg {
  type: "error";
  message: string;
}

export type ParticipantInboundMessage =
  | JoinMsg
  | LeaveMsg
  | OfferMsg
  | AnswerMsg
  | ICECandidateMsg;

export type ParticipantOutboundMessage =
  | WelcomeMsg
  | ParticipantViewMsg
  | ErrorMsg;

export type ParticipantMessage =
  | ParticipantInboundMessage
  | ParticipantOutboundMessage;

export interface HostStateMsg {
  type: "host-state";
  state: RoomState;
}

export interface ConfigMsg {
  type: "config";
  config: Partial<HostConfig>;
}

export interface ListCaptureDevicesMsg {
  type: "list-capture-devices";
}

export interface CaptureDevicesMsg {
  type: "capture-devices";
  devices: CaptureDevices;
}

export interface SetCaptureDevicesMsg {
  type: "set-capture-devices";
  selection: CaptureSelection;
}

export interface CaptureDevicesUpdatedMsg {
  type: "capture-devices-updated";
  capture: CaptureState;
  error?: string;
}

export interface ParticipantURLMsg {
  type: "participant-url";
  url: string;
}

export interface CopyParticipantURLMsg {
  type: "copy-participant-url";
}

export interface SetStreamProcessingMsg {
  type: "set-stream-processing";
  participantId: string;
  flags: StreamProcessingFlags;
}

export interface PreviewStreamInitMsg {
  type: "preview-stream-init";
  codec: string;
  width: number;
  height: number;
  fps: number;
}

export interface PreviewCutMsg {
  type: "preview-cut";
  activeVideoId: string;
  seq: number;
}

export type HostInboundMessage =
  | ConfigMsg
  | ListCaptureDevicesMsg
  | SetCaptureDevicesMsg
  | CopyParticipantURLMsg
  | SetStreamProcessingMsg;

export type HostOutboundMessage =
  | HostStateMsg
  | CaptureDevicesMsg
  | CaptureDevicesUpdatedMsg
  | ParticipantURLMsg
  | PreviewStreamInitMsg
  | PreviewCutMsg;

export type HostControlMessage = HostInboundMessage | HostOutboundMessage;

export type PreviewMessage = PreviewStreamInitMsg | PreviewCutMsg;

File: internal/protocol/messages.go
package protocol

type WelcomeMsg struct {
	Type     string              `json:"type"`
	ClientID string              `json:"clientId"`
	View     ParticipantRoomView `json:"view"`
}

type ParticipantViewMsg struct {
	Type string              `json:"type"`
	View ParticipantRoomView `json:"view"`
}

type JoinMsg struct {
	Type     string `json:"type"`
	Name     string `json:"name"`
	HasVideo bool   `json:"hasVideo"`
	HasAudio bool   `json:"hasAudio"`
	ClientID string `json:"clientId,omitempty"`
}

type LeaveMsg struct {
	Type string `json:"type"`
}

type OfferMsg struct {
	Type      string `json:"type"`
	From      string `json:"from"`
	To        string `json:"to"`
	SDP       string `json:"sdp"`
}

type AnswerMsg struct {
	Type string `json:"type"`
	From string `json:"from"`
	To   string `json:"to"`
	SDP  string `json:"sdp"`
}

type ICECandidateMsg struct {
	Type      string `json:"type"`
	From      string `json:"from"`
	To        string `json:"to"`
	Candidate string `json:"candidate"`
}

type ErrorMsg struct {
	Type    string `json:"type"`
	Message string `json:"message"`
}

type HostStateMsg struct {
	Type  string    `json:"type"`
	State RoomState `json:"state"`
}

type ConfigMsg struct {
	Type   string          `json:"type"`
	Config HostConfigPatch `json:"config"`
}

type ListCaptureDevicesMsg struct {
	Type string `json:"type"`
}

type CaptureDevicesMsg struct {
	Type    string         `json:"type"`
	Devices CaptureDevices `json:"devices"`
}

type SetCaptureDevicesMsg struct {
	Type      string           `json:"type"`
	Selection CaptureSelection `json:"selection"`
}

type CaptureDevicesUpdatedMsg struct {
	Type    string       `json:"type"`
	Capture CaptureState `json:"capture"`
	Error   string       `json:"error,omitempty"`
}

type ParticipantURLMsg struct {
	Type string `json:"type"`
	URL  string `json:"url"`
}

type CopyParticipantURLMsg struct {
	Type string `json:"type"`
}

type SetStreamProcessingMsg struct {
	Type          string                `json:"type"`
	ParticipantID string                `json:"participantId"`
	Flags         StreamProcessingFlags `json:"flags"`
}

type PreviewStreamInitMsg struct {
	Type   string `json:"type"`
	Codec  string `json:"codec"`
	Width  int    `json:"width"`
	Height int    `json:"height"`
	FPS    int    `json:"fps"`
}

type PreviewCutMsg struct {
	Type          string `json:"type"`
	ActiveVideoID string `json:"activeVideoId"`
	Seq           int    `json:"seq"`
}
